### PR TITLE
added test for training when both train and valid are subsets of a si…

### PR DESCRIPTION
**Pull request by j-mark-hou**
…ngle lgb.Dataset object.  

Added in response to a comment of @wxchan in #745

### CODE REVIEW

File: tests/python_package_test/test_engine.py [hunk 1]
```diff
@@ -453,3 +453,17 @@ def test_pandas_categorical(self):
...
         np.testing.assert_almost_equal(pred0, pred3)
         np.testing.assert_almost_equal(pred0, pred4)
+
+    def test_subset_train_val(self):
+        '''
+        Tests that it's fine to construct a single lgb.Dataframe object,
+        takes subsets of it, and uses the subsets for training and validation
+        '''
+        n = 1000
+        X = np.random.normal(size=n * 2).reshape(n, 2)
```
Comment (wxchan) [inline tests/python_package_test/test_engine.py]: you can use tuple `size = (n, 2)` to create 2-d array

File: tests/python_package_test/test_engine.py [hunk 2]
```diff
@@ -453,3 +453,17 @@ def test_pandas_categorical(self):
...
+        y = np.random.normal(size=n)
+        tmp_dat = lgb.Dataset(X, y)
+        tmp_dat_train = tmp_dat.subset(np.arange(int(n * .8)))
+        tmp_dat_val = tmp_dat.subset(np.arange(int(n * .8), n))
```
Comment (wxchan) [inline tests/python_package_test/test_engine.py]: Can you try subset of subset here? I think using your chain ref method, it will also work.

Comment (j-mark-hou) [inline tests/python_package_test/test_engine.py]: just added

File: tests/python_package_test/test_engine.py (hunk 1 above)
Comment (j-mark-hou) [inline tests/python_package_test/test_engine.py]: fixed

File: tests/python_package_test/test_engine.py (hunk 2 above)
Comment (wxchan) [inline tests/python_package_test/test_engine.py]: eh...you don't need to add another dataset, just make tmp_dat_val be subset of tmp_dat_train. You don't need to consider if the training make senses, we just want to test dataset constructing. You can simply use tmp_data_train.construct() to test rather than calling lgb.train().

Comment (j-mark-hou) [inline tests/python_package_test/test_engine.py]: This use case of using subsets in training is one that I explicitly care about, so I figured it would be worth having a test for this.  Would it be ok if I just did another PR with a test (probably in the basics.py test file) explicitly testing the ref_chain  logic?

Comment (wxchan) [inline tests/python_package_test/test_engine.py]: Doesn't this test case already covers ref_chain logic?

Comment (j-mark-hou) [inline tests/python_package_test/test_engine.py]: I think so.

I interpreted your last comment as suggesting I should replace the current test in this PR with one that explicitly tests the ref_chain logic instead of testing whether you can train a model with train / val as subsets.  My comment was to suggest that I felt like keeping this test was worthwhile, as I care about this use case, and the ref_chain logic is of purely instrumental value.  I suggested having a separate test chain logic as a compromise, but if you're fine with just this test then I'm happy to just keep thing as they are.

**Comment by guolinke** [comment]
@j-mark-hou @wxchan 
ready for merge ?

File: tests/python_package_test/test_engine.py (hunk 2 above)
Comment (wxchan) [inline tests/python_package_test/test_engine.py]: I meant you don't have to actually "train" it, just constructing dataset is fine. It will save some test time, as these tests are run after every submitted commit and every pr. Anyway, it's not a critial issue.

**Comment by wxchan** [comment]
@guolinke I think it's ok to merge.

**Comment by j-mark-hou** [comment]
go ahead yeah

**Comment by j-mark-hou** [comment]
actually, I think @wxchan has a point I didn't realize the test time was a major constraint.  I'll fix this tonight and let's merge then instead.

**Comment by wxchan** [comment]
@j-mark-hou actually the time increase is trivial. You can remove some duplicate cases. One subset and one subset of subset is good.

**Comment by j-mark-hou** [comment]
alright i'll do that tonight thanks

**Comment by wxchan** [comment]
@guolinke ready to merge.
